gluten-core/src/main/scala/org/apache/gluten/extension/columnar/heuristic/HeuristicApplier.scala

```diff
@@ -0,0 +1,173 @@
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.rules.{PlanChangeLogger, Rule}
+import org.apache.spark.sql.execution.{ColumnarCollapseTransformStages, GlutenFallbackReporter, SparkPlan}
+import org.apache.spark.util.SparkRuleUtil
+
+class HeuristicApplier(session: SparkSession)
+  extends ColumnarRuleApplier
+  with Logging
+  with LogLevelUtil {
+
+  private lazy val transformPlanLogLevel = GlutenConfig.getConf.transformPlanLogLevel
+  private lazy val planChangeLogger = new PlanChangeLogger[SparkPlan]()
+
+  private val adaptiveContext = AdaptiveContext(session)
+
+  override def apply(plan: SparkPlan, outputsColumnar: Boolean): SparkPlan =
+    withTransformRules(transformRules(outputsColumnar)).apply(plan)
+
+  // Visible for testing.
+  def withTransformRules(transformRules: List[SparkSession => Rule[SparkPlan]]): Rule[SparkPlan] =
+    plan =>
+      PhysicalPlanSelector.maybe(session, plan) {
+        val finalPlan = prepareFallback(plan) {
+          p =>
+            val suggestedPlan = transformPlan(transformRules, p, "transform")
+            transformPlan(fallbackPolicies(), suggestedPlan, "fallback") match {
+              case FallbackNode(fallbackPlan) =>
+                // we should use vanilla c2r rather than native c2r,
+                // and there should be no `GlutenPlan` any more,
+                // so skip the `postRules()`.
+                fallbackPlan
+              case plan =>
+                transformPlan(postRules(), plan, "post")
+            }
+        }
+        transformPlan(finalRules(), finalPlan, "final")
+      }
+
+  private def transformPlan(
+      getRules: List[SparkSession => Rule[SparkPlan]],
+      plan: SparkPlan,
+      step: String) = GlutenTimeMetric.withMillisTime {
+    logOnLevel(
+      transformPlanLogLevel,
+      s"${step}ColumnarTransitions preOverridden plan:\n${plan.toString}")
+    val overridden = getRules.foldLeft(plan) {
+      (p, getRule) =>
+        val rule = getRule(session)
+        val newPlan = rule(p)
+        planChangeLogger.logRule(rule.ruleName, p, newPlan)
+        newPlan
+    }
+    logOnLevel(
+      transformPlanLogLevel,
+      s"${step}ColumnarTransitions afterOverridden plan:\n${overridden.toString}")
+    overridden
+  }(t => logOnLevel(transformPlanLogLevel, s"${step}Transform SparkPlan took: $t ms."))
+
+  private def prepareFallback[T](plan: SparkPlan)(f: SparkPlan => T): T = {
+    adaptiveContext.setAdaptiveContext()
+    adaptiveContext.setOriginalPlan(plan)
+    try {
+      f(plan)
+    } finally {
+      adaptiveContext.resetOriginalPlan()
+      adaptiveContext.resetAdaptiveContext()
+    }
+  }
+
+  /**
+   * Rules to let planner create a suggested Gluten plan being sent to `fallbackPolicies` in which
+   * the plan will be breakdown and decided to be fallen back or not.
+   */
+  private def transformRules(outputsColumnar: Boolean): List[SparkSession => Rule[SparkPlan]] = {
```

Comment:
Looks there are few difference between the rule sets (`transformRules`, `fallbackPolicies`, `postRules`, `finalRules`) of `HeuristicApplier` & `EnumeratedApplier`. Will some more differences be introduced in the future development? If not, may be better to put common rules in some common place or super class?

Reply:
Actually it's intentional to have some duplicated code in `HeuristicApplier` and `EnumeratedApplier` for this PR. In my plan `EnumeratedApplier`'s code will be frequently updated in the future PRs of RAS, and end up with very different design with `HeuristicApplier` so the two will end up sharing less (or event no) code. 

So the duplicated code is for better isolation of the two implementations to ease future development.

Reply:
@zhztheplayer, I see. Thanks for the clarification!